python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -0,0 +1,311 @@
+                except StopIteration:
+                    break
+                batch_buffer.append(block_ref)
+                buffer_size += metadata.num_rows
+                num_rows_in_last_block = metadata.num_rows
+
+            if buffer_size == batch_size:
+                # If equal to batch size, then yield the full buffer.
+                yield LogicalBatch(
+                    global_index, batch_buffer, starting_index, None, buffer_size
+                )
+                batch_buffer = []
+                buffer_size = 0
+                starting_index = 0
+                num_rows_in_last_block = 0
+                global_index += 1
+
+            if buffer_size > batch_size:
+                # If current buffer is greater than batch size, then yield part of the
+                # buffer, and carryover the remainder to the next batch.
+                num_rows_to_leave_behind = buffer_size - batch_size
+                ending_index = num_rows_in_last_block - num_rows_to_leave_behind
+                assert ending_index > 0, ending_index
+                yield LogicalBatch(
+                    global_index, batch_buffer, starting_index, ending_index, batch_size
+                )
+                global_index += 1
+                # Carryover to next batch.
+                batch_buffer = [batch_buffer[-1]]
+                buffer_size = num_rows_to_leave_behind
+                starting_index = ending_index
+
+    # Yield any leftover batches if necessary.
+    if buffer_size > 0 and not drop_last:
+        assert buffer_size < batch_size
+        yield LogicalBatch(
+            global_index, batch_buffer, starting_index, None, buffer_size
+        )
+        global_index += 1
+
+
+def local_shuffle_logical_batches(
+    logical_batch_iterator: Iterator[LogicalBatch],
+    shuffle_buffer_min_size: int,
+    shuffle_seed: Optional[int] = None,
+) -> Iterator[LogicalBatch]:
+    """Shuffles the logical batch iterator using a buffer of the provided size."""
+
+    if shuffle_seed is not None:
+        random.seed(shuffle_seed)
+
+    shuffle_buffer: List[LogicalBatch] = []
+    shuffle_buffer_size = 0
+    global_counter = 0
+
+    for logical_batch in logical_batch_iterator:
+        shuffle_buffer.append(logical_batch)
+        shuffle_buffer_size += logical_batch.num_rows
+
+        while shuffle_buffer_size >= shuffle_buffer_min_size:
+            output_batch = shuffle_buffer.pop(
+                random.randint(0, len(shuffle_buffer) - 1)
+            )
+            output_batch.batch_idx = global_counter
+            yield output_batch
+            shuffle_buffer_size -= output_batch.num_rows
+            global_counter += 1
+
+    # Yield any leftover.
+    while len(shuffle_buffer) > 0:
+        output_batch = shuffle_buffer.pop(random.randint(0, len(shuffle_buffer) - 1))
+        output_batch.batch_idx = global_counter
+        yield output_batch
+        global_counter += 1
+
+
+def prefetch_batches_locally(
```

Comment:
Why don't we consolidate prefetching via the `BlockPrefetcher` with resolving blocks in parallel over multiple threads via `ray.get()`? I.e. we don't need both `prefetch_batches` and `prefetch_blocks` at the API level, and we don't need `BlockPrefetcher`/`prefetch_batches_locally`, when we can resolve (call `ray.get()`) concurrently across `prefetch_batches` threads. `prefetch_batches` would completely control the prefetching behavior, and we'd use the `batch_size` + the block metadata to know how many blocks to resolve in each thread (this is already taken care of via the bundling to create `LogicalBatch`s).

Reply:
No longer applicable since we decided to do prefetching and `ray.get()` in a single thread.